compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNotImplementedOverrideChecker.kt

```diff
@@ -0,0 +1,278 @@
+                result.add(originalSymbol.fir)
+            }
+        }
+
+        fun checkPropertySymbolAndAddToResult(originalSymbol: FirCallableSymbol<*>) {
+            if (shouldCreateFakeOverridden<FirProperty, FirPropertySymbol>(firClass, originalSymbol, contributedDeclarations)) {
+                result.add(originalSymbol.fir)
+            }
+        }
+
+        val superTypesCallableNames = classScope.getCallableNames()
+        for (name in superTypesCallableNames) {
+            classScope.processFunctionsByName(name) { functionSymbol ->
+                // TODO: MANY_* as well as some conflict diagnostics
+                //if (functionSymbol is FirIntersectionOverrideFunctionSymbol)
+                //    functionSymbol.intersections.forEach(::checkFunctionSymbolAndAddToResult)
+                //else
+                checkFunctionSymbolAndAddToResult(functionSymbol)
+            }
+
+            classScope.processPropertiesByName(name) { propertySymbol ->
+                // TODO: MANY_* as well as some conflict diagnostics
+                //if (propertySymbol is FirIntersectionOverridePropertySymbol)
+                //    propertySymbol.intersections.forEach(::checkPropertySymbolAndAddToResult)
+                //else
+                checkPropertySymbolAndAddToResult(propertySymbol)
+            }
+        }
+
+        return result
+    }
+
+    // See [FakeOverrideGenerator#createFakeOverriddenIfNeeded]
+    private inline fun <reified D : FirCallableMemberDeclaration<D>, reified S : FirCallableSymbol<D>> shouldCreateFakeOverridden(
+        firClass: FirClass<*>,
+        originalSymbol: FirCallableSymbol<*>,
+        contributedDeclarations: Collection<FirDeclaration>,
+    ): Boolean {
+        if (originalSymbol !is S || originalSymbol.fir in contributedDeclarations) return false
+        val classLookupTag = firClass.symbol.toLookupTag()
+        val originalDeclaration = originalSymbol.fir
+        if (originalSymbol.dispatchReceiverClassOrNull() == classLookupTag && !originalDeclaration.origin.fromSupertypes) return false
+        if (originalDeclaration.visibility == Visibilities.Private) return false
+        return when {
+            originalSymbol.fir.origin.fromSupertypes && originalSymbol.dispatchReceiverClassOrNull() == classLookupTag -> {
+                // Substitution case
+                false
+            }
+            originalDeclaration.allowsToHaveFakeOverrideIn(firClass) -> {
+                // Trivial fake override case
+                true
+            }
+            else -> {
+                false
+            }
+        }
+    }
+
+    // TODO: Need to refactor the counterpart in FakeOverrideGenerator?
+    private fun FirCallableMemberDeclaration<*>.allowsToHaveFakeOverrideIn(firClass: FirClass<*>): Boolean {
+        if (!allowsToHaveFakeOverride) return false
+        // if (this.visibility != JavaDescriptorVisibilities.PACKAGE_VISIBILITY) return true
```

Comment:
This is very questionable place. Why should you check package equality in case of non-package private declaration? I understand that Java visibility isn't accessible at this point, but nevertheless we should implement the necessary logic in some way. I will think about this place.

Reply:
Please look at `testIrrelevantImplCharSequenceKotlin` from old FE tests. Looks like you do not handle multi-inheritance properly, and this test works in your branch more or less accidentally.

Reply:
I think I found a nicer way: using `FirVisibilityChecker`'s `isVisible`. No behavior changes!

Reply:
> Looks like you do not handle multi-inheritance properly

I have to admit that's the case. I left a bunch of TODOs regarding that. This visibility check seems a different topic, though.